drivers/modem/gsm_ppp.c

```diff
@@ -669,6 +719,49 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		return;
 	}
 
+registering:
+	/* Wait for cell tower registration */
+	ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &check_net_reg_cmd, 1,
+				    "AT+CREG?",
+				    &gsm->sem_response,
+				    GSM_CMD_SETUP_TIMEOUT);
+	if ((ret < 0) || ((gsm->net_state != GSM_ROAMING) &&
+			 (gsm->net_state != GSM_HOME_NETWORK))) {
+		if (!gsm->register_retries) {
+			gsm->register_retries = CONFIG_MODEM_GSM_REGISTER_TIMEOUT *
+				MSEC_PER_SEC / GSM_REGISTER_DELAY_MSEC;
+		} else {
+			gsm->register_retries--;
+
+			/* Reset RF if timed out */
+			if (!gsm->register_retries) {
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
+							    &gsm->context.cmd_handler,
+							    &response_cmds[0],
+							    ARRAY_SIZE(response_cmds),
+							    "AT+CFUN=0", &gsm->sem_response,
+							    GSM_CMD_AT_TIMEOUT);
```

Comment:
Is this really needed? In my experience it is better to let the modem do its own thing and let it keep searching instead of interrupting its process by turning off the radio. This should be a last resort after giving the modem minutes (~10) to try and get on a network.

Reply:
This was suggested by @WilliamGFish and is used in the ublox sara driver. Personally I'm not sure if this is necessary.

Reply:
I agree that it's better to just let it spin. This way the application has better control over when to give up and try again.

Reply:
@rerickson1 @endian-benjamin I agree with you all